### ISSUE TITLE
Concurrency safe write does bot include hostname in key to hash

### ISSUE THREAD

**Issue by LunarLanding** [joblib/joblib#1610]
Should this code include the hostname in the temporary filename key? Otherwise processes in different nodes using a shared file system could end up writing to the same file, no?

https://github.com/joblib/joblib/blob/bca1f4216a38cff82a85371c45dde79bed977d0e/joblib/_store_backends.py#L31-L40

**Comment by lesteve** [comment]
There is `os.getpid()` in the filename, so processes of different nodes would get a different pid, unless I am missing something?

**Comment by LunarLanding** [comment]
@lesteve pids are not unique across different nodes. Process 100 in nodeA and process 100 in nodeB could end up wiring to the same location.

**Comment by lesteve** [comment]
Indeed, let's reopen this! Did it actually pose problems in real life or is your issue based on reading the source code?

**Comment by LunarLanding** [comment]
> Indeed, let's reopen this! Did it actually pose problems in real life or is your issue based on reading the source code?

It's been a while, so I am not exactly sure. I definitely had issues in the past with using libraries that do not account for shared usage on a NFS filesystem over multiple nodes. I compared joblib's implementation to python's afaik best standard for getting a unique filename in this situation in [flufl.lock](https://gitlab.com/warsaw/flufl.lock/-/blob/main/src/flufl/lock/_lockfile.py?ref_type=heads#L471), and found the hostname was not included, henceforth this issue.

**Comment by lesteve** [comment]
OK thanks a lot for the additional info!

We could improve the situation on this particular line, but I would not be surprised if there are other issues when using NFS filesystems, because this is the kind of things we are not thoroughly testing.

I seem to vaguely recall issues in the past that had to do with NFS quirks, where some things are not atomic and another maintainer trying to read the NFS spec to figure out what to do. I don't remember what came out of it though 😅.

**Comment by LunarLanding** [comment]
I think that adding the hostname to the filename will at least fix the use case where many actors over multiple nodes create cache entries. Cache deletion on the other hand needs to be done on a single machine. This matches mine and my colleagues' use case.
The flufl.lock repo has some detail on how to create a unique file on [NFS](https://gitlab.com/warsaw/flufl.lock/-/blob/main/src/flufl/lock/_lockfile.py?ref_type=heads#L111):
```
This class implements an NFS-safe file-based locking algorithm
    influenced by the GNU/Linux open(2) manpage, under the description
    of the O_EXCL option:

        [...] O_EXCL is broken on NFS file systems, programs which rely on it
        for performing locking tasks will contain a race condition.  The
        solution for performing atomic file locking using a lockfile is to
        create a unique file on the same fs (e.g., incorporating hostname and
        pid), use link(2) to make a link to the lockfile.  If link() returns
        0, the lock is successful.  Otherwise, use stat(2) on the unique file
        to check if its link count has increased to 2, in which case the lock
        is also successful.

```
